vhost-device-spi/src/spi.rs

```diff
@@ -0,0 +1,1055 @@
+// Low level SPI definitions
+//
+// Copyright (c) 2024 Qualcomm Innovation Center, Inc. All rights reserved.
+//          Haixu Cui <quic_haixcui@quicinc.com>
+//
+// SPDX-License-Identifier: Apache-2.0 or BSD-3-Clause
+
+use std::{
+    convert::From,
+    fs::{File, OpenOptions},
+    os::unix::io::AsRawFd,
+    ptr,
+};
+
+use thiserror::Error as ThisError;
+use vmm_sys_util::errno::Error as IoError;
+use vmm_sys_util::ioctl::{ioctl_with_mut_ptr, ioctl_with_mut_ref, ioctl_with_ref};
+
+use crate::{vhu_spi::VirtioSpiConfig, virtio_spi::*};
+
+type Result<T> = std::result::Result<T, Error>;
+
+#[derive(Copy, Clone, Debug, PartialEq, ThisError)]
+/// Errors related to low level spi helpers
+pub(crate) enum Error {
+    #[error("Ioctl command failed for {0} operation: {1}")]
+    IoctlFailure(&'static str, IoError),
+    #[error("Failed to open spi controller")]
+    DeviceOpenFailed,
+}
+
+/// Describes a single SPI transfer
+#[derive(Debug)]
+#[repr(C)]
+pub(crate) struct SpiIocTransfer {
```

Comment:
Where did this structure come from? If it's defined in a C header, would it be better to use bindgen to generate them so they can be updated semi-automatically in the future?

Reply:
If it's just a structure and we don't want to use bindgen that's fine, but at least write a comment explaining where it came from. I think I found it in Linux's UAPI (`include/uapi/linux/spi/spidev.h`).

Reply:
And maybe we should move it in the same file where we define the IOCTLs.

Reply:
We could the generation on the build.rs file, perhaps. Also I agree that moving all C bindings to separate modules is better to make the distinction clear.

Reply:
@epilys don't know if generating the files every time we compile has risks (e.g. a minimal container used only for compilation, may not have the kernel headers installed). For example in https://github.com/rust-vmm/kvm-bindings we already commit them generated, so if we want to use `bindgen`, maybe I would already commit the final result.

Reply:
Yes, this structure is in Linux `include/uapi/linux/spi/spidev.h`, and I will move it to `virtio_spi.rs`.

Reply:
Agreed with all ++